vcr/issuer/openid4vci/store.go

```diff
@@ -0,0 +1,194 @@
+	mux      *sync.RWMutex
+	flows    map[string]Flow
+	refs     map[string]map[string]referenceValue
+	routines *sync.WaitGroup
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+// NewMemoryStore creates a new in-memory Store.
+func NewMemoryStore() Store {
+	result := &memoryStore{
+		mux:      &sync.RWMutex{},
+		flows:    map[string]Flow{},
+		refs:     map[string]map[string]referenceValue{},
+		routines: &sync.WaitGroup{},
+	}
+	result.startPruning(pruneInterval)
+	return result
+}
+
+type referenceValue struct {
+	FlowID string    `json:"flow_id"`
+	Expiry time.Time `json:"exp"`
+}
+
+func (o *memoryStore) Store(_ context.Context, flow Flow) error {
+	if len(flow.ID) == 0 {
+		return errors.New("invalid flow ID")
+	}
+	o.mux.Lock()
+	defer o.mux.Unlock()
+	if o.flows[flow.ID].ID != "" {
+		return errors.New("OAuth2 flow with this ID already exists")
+	}
+	o.flows[flow.ID] = flow
+	return nil
+}
+
+func (o *memoryStore) StoreReference(_ context.Context, flowID string, refType string, reference string, expiry time.Time) error {
+	if len(reference) == 0 {
+		return errors.New("invalid reference")
+	}
+	o.mux.Lock()
+	defer o.mux.Unlock()
+	if o.flows[flowID].ID == "" {
+		return errors.New("OAuth2 flow with this ID does not exist")
+	}
+	if o.refs[refType] == nil {
+		o.refs[refType] = map[string]referenceValue{}
+	}
+	if _, ok := o.refs[refType][reference]; ok {
+		return errors.New("reference already exists")
+	}
+	o.refs[refType][reference] = referenceValue{FlowID: flowID, Expiry: expiry}
+	return nil
+}
+
+func (o *memoryStore) FindByReference(_ context.Context, refType string, reference string) (*Flow, error) {
+	o.mux.RLock()
+	defer o.mux.RUnlock()
+
+	refMap := o.refs[refType]
+	if refMap == nil {
+		return nil, nil
+	}
+	value, ok := refMap[reference]
+	if !ok {
+		return nil, nil
+	}
+	if value.Expiry.Before(time.Now()) {
+		return nil, nil
+	}
+
+	flow := o.flows[value.FlowID]
+	if flow.Expiry.Before(time.Now()) {
+		return nil, nil
+	}
+	return &flow, nil
+}
+
+func (o *memoryStore) DeleteReference(_ context.Context, refType string, reference string) error {
+	o.mux.Lock()
+	defer o.mux.Unlock()
+
+	if o.refs[refType] == nil {
+		return nil
+	}
+	delete(o.refs[refType], reference)
+	return nil
+}
+
+func (o *memoryStore) Close() {
+	// Signal pruner to stop and wait for it to finish
+	o.cancel()
+	o.routines.Wait()
+}
+
+func (o *memoryStore) startPruning(interval time.Duration) {
+	o.ctx, o.cancel = context.WithCancel(context.Background())
+	ticker := time.NewTicker(interval)
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				flowsPruned, refsPruned := o.prune(time.Now())
+				if flowsPruned > 0 || refsPruned > 0 {
+					log.Logger().Debugf("Pruned %d expired OpenID4VCI flows and %d expired refs", flowsPruned, refsPruned)
+				}
+			}
+		}
+	}(o.ctx)
+}
+
+func (o *memoryStore) prune(moment time.Time) (int, int) {
```

Comment:
In all cases this method is called with `time.Now()` as argument. Why not use time.Now within the method?